source/plugin/event/causes.rst

```diff
@@ -69,55 +97,137 @@ object matching the provided type.
+    CommandSource sourceRunningSudo;
+    Player playerToSimulate;
+    try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+      // This would be the player that is causing the action (would have run the /sudo command)
+      frame.pushCause(sourceRunningSudo);
+
+      // You may want to push the simulated player so that it's at the root for compatibility
+      // with other plugins that listen for the direct cause
+      frame.pushCause(playerToSimulate);
+
+      // This would contain the GameProfile of the player to simulate, they are not a direct
+      // cause, but provide additional information to the event listeners so that they know
+      // the root player is a simulated one and may not have invoked the event.
+      frame.addContext(EventContextKeys.SIMULATED_PLAYER, playerToSimulate.getProfile());
+
+      // Add your event code here.
+      // You can get the current cause at any time using the getCurrentCause method
+      // (which is available on the fram as well as the CauseStackManager)
+      Cause cause = frame.getCurrentCause();
+    }
+
+Note that the last item you push to the cause stack will be the root of the ``Cause``.
+
+Using the Cause Builder
+=======================
+
+If you are creating an event that does not fire on the main thread, you cannot use the
+``CauseStackManager``. Instead, you will need to create a ``Cause`` object manually.
+
+Creating a cause object is easy using the :javadoc:`Cause.Builder`. You can obtain a
+builder by calling ``Cause.builder()``. To add a cause to the builder, use the
+:javadoc:`Cause.Builder#append(Object)` method, but note that unlike the ``CauseStackManager``,
+the first element you add will be the root, not the last.
 
-If you already have a cause object and would like to append some more objects to the chain you can use
-:javadoc:`Cause#with(NamedCause, NamedCause...)`. This constructs a new Cause object containing first the objects
-already present in the original cause, then followed by the additional objects that you provided.
+If you wish to add contexts, there is a separate builder for those, the
+:javadoc:`EventContext.Builder`, accessed by calling ``EventContext#builder()``.
+The ``EventContext`` can then be added using the ``Cause.Builder#build(EventContext)`` when
+you have finished building the ``Cause`` up.
+
+Taking the previous example, this is how we would build it using the cause builder.
+
+**Creating a custom Cause with the Cause and EventContext builders**
+
+.. code-block:: java
+
+    // In your code these would be populated
+    CommandSource sourceRunningSudo;
+    Player playerToSimulate;
+    PluginContainer plugin;
+
+    Cause cause = Cause.builder()
+      .append(playerToSimulate) // This would be root
+      .append(sourceRunningSudo)
+      .append(plugin)
+      .build(EventContext.builder().add(EventContextKeys.PLAYER_SIMULATED, playerToSimulate.getProfile()).build());
```

Comment:
IMO this should be formatted differently.
Either move it to its own variable or apply the same line breaks as above.